include/pybind11/cast.h

```diff
@@ -686,13 +694,32 @@ class type_caster_generic {
     }
     void check_holder_compat() {}
 
+    /// Try to load with foreign typeinfo, if available. Used when there is no
+    /// native typeinfo, or when the native one wasn't able to produce a value.
+    PYBIND11_NOINLINE bool try_load_foreign_module_local(handle src) {
+        constexpr auto *local_key = "_pybind11_module_local_typeinfo";
+        const auto pytype = src.get_type();
+        if (!hasattr(pytype, local_key))
+            return false;
+
+        type_info *foreign_typeinfo = reinterpret_borrow<capsule>(getattr(pytype, local_key));
+        if (foreign_typeinfo == typeinfo)
+            return false; // Only consider this foreign loader if actually foreign
```

Comment:
Do you mean a local binding masking a local one? That should work and is covered by the tests. (Not the `cpptype` mismatch mentioned below.)